### ISSUE TITLE
Simple Hello World Fails

### ISSUE THREAD

**Issue by ppey** [lunixbochs/usercorn#230]
When I run usercon with a simple hello wold c application, usercorn fails with the following output. Any suggestion what went wrong ? I'm using an updated arch Linux on amd64 and the helloworld is compiled using default gcc settings.

 ./main hello-world 
invalid read: @0x0, 0x1 = 0x0
[memory map]
  0x60000000-0x60800000 rwx [stack]
  0x60800000-0x60801000 ---
  0x400000-0x401000 r-x hello-world [exe]
  0x600000-0x602000 rwx hello-world [exe]
  0x1e02000-0x2027000 rwx ld-linux-x86-64.so.2 [interp]
  0x100000-0x128000 r-- ld.so.cache
  0x623000-0x99c000 rwx libc.so.6
  0x99c000-0x9a2000 rwx libc.so.6
  0x9a2000-0x9a6000 rw-
  0x128000-0x12a000 rw-
  0x602000-0x623000 rwx [brk]
[registers]
  rax 0x0000000000000000   rdi 0x000000000060201b   rsp 0x00000000607ff750   r11 0x0000000000000000 
  rbp 0x0000000000400510   rdx 0x00000000009a2740    r8 0x0000000000602000   r12 0x0000000000400400 
  rbx 0x0000000000000000   rip 0x0000000000622511    r9 0x000000000000000b   r13 0x00000000607ff820 
  rcx 0x0000000000000000   rsi 0x0000000000000000   r10 0x00000000009a0b38   r14 0x0000000000000000 
                                                                             r15 0x0000000000000000 
[stacktrace]
  0x622511 __libc_start_main+0xf1
panic: Invalid memory read (UC_ERR_READ_UNMAPPED)

goroutine 1 [running]:
github.com/lunixbochs/usercorn/go/cmd.(*UsercornCmd).Run(0xc420016690, 0xc4200189c0, 0x4, 0x4, 0xc42002bc00, 0x2b, 0x40)
	~/go/src/github.com/lunixbochs/usercorn/go/cmd/cmd.go:290 +0x1422
main.main()
	~/go/src/github.com/lunixbochs/usercorn/go/cmd/usercorn/main.go:12 +0x7f

**Comment by ellenpoe** [comment]
I've been having the same problem with my hello world attempt. I also get similar output ("ld-linux-x86-64.so.2 [interp]" in the stack trace) when trying to run any of the simple system binaries I've tried (cat, echo, pwd).

I tried to test v0.2 but I'm having trouble building the dependencies. I'll update with any useful info as I get it.

**Comment by lunixbochs** [comment]
If you build 0.2 you should use the makefile from 0.3. I'll just cut a 0.2.1 tag.

**Comment by lunixbochs** [comment]
v0.2.1 is cut, but it looks like it has the same problem

Something like /bin/ls partially works for me, but hello world does nothing.

**Comment by ellenpoe** [comment]
I also have the problem on v0.2 with the v0.3 makefile copied over. I'll try v0.2.1 and master (again) when I get a chance. Is there documentation for how to try usercorn with a different userspace so I can eliminate that as a possible problem?

**Comment by lunixbochs** [comment]
`usercorn -prefix <other userspace root>`

**Comment by ellenpoe** [comment]
Perfect, thank you. I'll update later today if I find anything useful.

**Comment by lunixbochs** [comment]
With arch, my hello world *works*, but crashes on exit with a similar error to @ppey. Is your hello world empty? `__libc_start_main+0xf1` looks like the exit or .fini position.

```
$ ./usercorn -prefix linux-arch64/ ./a.out
hello world
invalid read: @0xc, 0x1 = 0x0
[memory map]
  0x60000000-0x60800000 rwx [stack]
  0x60800000-0x60801000 ---
  0x400000-0x401000 r-x a.out [exe]
  0x600000-0x602000 rwx a.out [exe]
  0x1e02000-0x2027000 rwx ld-2.25.so [interp]
  0x100000-0x106000 r-- ld.so.cache
  0x625000-0x99c000 rwx libc-2.25.so
  0x99c000-0x9a2000 rwx libc-2.25.so
  0x9a2000-0x9a6000 rw-
  0x106000-0x108000 rw-
  0x602000-0x625000 rwx [brk]
[registers]
  rax 0x000000000000000c   rdi 0x0000000000000000   rsp 0x00000000607ff970   r11 0x0000000000000000 
  rbp 0x0000000000400540   rdx 0x00000000009a2740    r8 0x0000000000602000   r12 0x0000000000400440 
  rbx 0x0000000000000000   rip 0x0000000000622511    r9 0x0000000000000000   r13 0x00000000607ffa40 
  rcx 0x00000000fbad2a84   rsi 0x0000000000602010   r10 0x00000000009a0b38   r14 0x0000000000000000 
                                                                             r15 0x0000000000000000 
[stacktrace]
  0x622511 __libc_start_main+0xf1
panic: Invalid memory read (UC_ERR_READ_UNMAPPED)
```

**Comment by ellenpoe** [comment]
I'm not sure what @ppey is using for their hello world, but when I printf something not containing a newline, I do not get any output other than the crash information.

The following code:
```
#include <stdio.h>

int main() {
  printf("hello world!\n");
  return 0;
}
```
yields the following output

```
ellen@ellen-Gandof:~/dev/usercorn$ ./usercorn ../basic/a.out
hello world!
invalid read: @0x0, 0x1 = 0x0
[memory map]
  0x60000000-0x60800000 rwx [stack]
  0x60800000-0x60801000 ---
  0x400000-0x401000 r-x a.out [exe]
  0x600000-0x602000 rwx a.out [exe]
  0x1e02000-0x202a000 rwx ld-linux-x86-64.so.2 [interp]
  0x100000-0x102000 rw-
  0x102000-0x118000 r-- ld.so.cache
  0x623000-0x9c1000 rwx libc.so.6
  0x9c1000-0x9c7000 rwx libc.so.6
  0x9c7000-0x9cb000 rw-
  0x118000-0x119000 rw-
  0x119000-0x11a000 rw-
  0x602000-0x623000 rwx [brk]
[registers]
  rax 0x0000000000000000   rdi 0x0000000000000001   rsp 0x00000000607feea0   r11 0x0000000000000000 
  rbp 0x0000000000400540   rdx 0x00000000009c7780    r8 0x0000000000602000   r12 0x0000000000400430 
  rbx 0x0000000000000000   rip 0x0000000000622830    r9 0x000000000000000d   r13 0x00000000607fef70 
  rcx 0x00000000fbad2a84   rsi 0x0000000000602010   r10 0x00000000009c5b78   r14 0x0000000000000000 
                                                                             r15 0x0000000000000000 
[stacktrace]
  0x622830
panic: Invalid memory read (UC_ERR_READ_UNMAPPED)

goroutine 1 [running]:
github.com/lunixbochs/usercorn/go/cmd.(*UsercornCmd).Run(0xc420018640, 0xc42000c420, 0x2, 0x2, 0xc4200f2800, 0x40, 0x40)
	/home/ellen/dev/usercorn/.gopath/src/github.com/lunixbochs/usercorn/go/cmd/cmd.go:286 +0x1436
main.main()
	/home/ellen/dev/usercorn/go/cmd/usercorn/main.go:12 +0x7f
```

It looks like programs that do not make any system calls behave correctly. A busy wait works, for example. A printf("output\n") followed by a busy wait causes a crash.

I'm running ubuntu 16.04 so maybe my glibc is more recent than yours on 14.04? Mine is 2.23.

**Comment by lunixbochs** [comment]
stdout in glibc is line-buffered. printf failing without a newline means the libc crashed before finalizers executed (which includes an fflush(stdout))

**Comment by ellenpoe** [comment]
I tried calling fflush(stdout) manually, with interesting results.
```
#include <stdio.h>

int main() {
  printf("output!\n");
  fflush(stdout);
}
```

Produces the following output:

```
ellen@ellen-Gandof:~/dev/usercorn$ ./usercorn ../basic/a.out
output!
../basic/a.out: relocation error: ../basic/a.out: symbol fflush, version GLIBC_2.2.5 not defined in file libc.so.6 with link time reference
```

I don't really know what to make of this. I'll keep digging though.

**Comment by lunixbochs** [comment]
Fixed by 1285b5c (latest unstable)

**Comment by ppey** [comment]
Thanks, now everything works just fine!
